pennylane/gradients/hadamard_gradient.py

```diff
@@ -309,218 +290,122 @@ def hadamard_grad(
+            new_val = [qml.math.convert_like(2 * coeff * res, res)]
+        final_res.append(new_val)
+
+    # Post process for probs
     measurements_probs = [
         idx
         for idx, m in enumerate(tape.measurements)
         if isinstance(m, qml.measurements.ProbabilityMP)
     ]
-
-    def _postprocess_probs(res, measurement, projector):
-        num_wires_probs = len(measurement.wires)
-        if num_wires_probs == 0:
-            num_wires_probs = tape.num_wires
-        res = qml.math.reshape(res, (2**num_wires_probs, 2))
-        return qml.math.tensordot(res, projector, axes=[[1], [0]])
-
-    def processing_fn(results):  # pylint: disable=too-many-branches
-        """Post processing function for computing a hadamard gradient."""
-        final_res = []
-        for coeff, res in zip(coeffs, results):
-            if isinstance(res, tuple):
-                new_val = [qml.math.convert_like(2 * coeff * r, r) for r in res]
-            else:
-                new_val = qml.math.convert_like(2 * coeff * res, res)
-            final_res.append(new_val)
-
-        # Post process for probs
-        if measurements_probs:
-            projector = np.array([1, -1])
-            like = final_res[0][0] if multi_measurements else final_res[0]
-            projector = qml.math.convert_like(projector, like)
-            for idx, res in enumerate(final_res):
-                if multi_measurements:
-                    for prob_idx in measurements_probs:
-                        final_res[idx][prob_idx] = _postprocess_probs(
-                            res[prob_idx], tape.measurements[prob_idx], projector
-                        )
-                else:
-                    prob_idx = measurements_probs[0]
-                    final_res[idx] = _postprocess_probs(res, tape.measurements[prob_idx], projector)
-        grads = []
-        idx = 0
-        for num_tape in gradient_data:
-            if num_tape == 0:
-                grads.append(qml.math.zeros(()))
-            elif num_tape == 1:
-                grads.append(final_res[idx])
-                idx += 1
-            else:
-                result = final_res[idx : idx + num_tape]
-                if multi_measurements:
-                    grads.append(
-                        [qml.math.array(qml.math.sum(res, axis=0)) for res in zip(*result)]
-                    )
-                else:
-                    grads.append(qml.math.array(qml.math.sum(result)))
-                idx += num_tape
-
-        if not multi_measurements and not multi_params:
-            return grads[0]
-
-        if not (multi_params and multi_measurements):
-            if multi_measurements:
-                return tuple(grads[0])
-            return tuple(grads)
-
-        # Reordering to match the right shape for multiple measurements
-        grads_reorder = [[0] * len(tape.trainable_params) for _ in range(len(tape.measurements))]
-
-        for i in range(len(tape.measurements)):
```

Comment:
By preparing the data structure in the correct shape, we don't need a follow on step to reshape everything.